packages/core/src/lib/lit-core.ts

```diff
@@ -637,16 +647,42 @@ export class LitCore {
     const serverKeys: Record<string, JsonHandshakeResponse> = {};
 
     let timeoutHandle: ReturnType<typeof setTimeout>;
+    let canRetry = this.config.retryNodeHandshake ?? true;
+
+    const handshakeWithNode = async (url: string): Promise<void> => {
+      serverKeys[url] = await this._handshakeAndVerifyNodeAttestation({
+        url,
+        requestId,
+      });
+      connectedNodes.add(url);
+    };
+
+    const handshakeWithNodeWithRetry = async (url: string): Promise<void> => {
+      try {
+        return await handshakeWithNode(url);
+      } catch (e) {
+        if (!canRetry) {
+          throw e;
+        }
+      }
+
+      // If we get here, we've failed to handshake with the node. We should retry.
+      logErrorWithRequestId(
+        requestId,
+        `Error while attempting handshake with node ${url}. Retrying...`
+      );
+      return await handshakeWithNodeWithRetry(url);
+    };
```

Comment:
I think we need to add a max retry here for the recursion to avoid stack overflow if the number of retries is large, and it looks like the canRetry flag remains unchanged if the handshake fails?

```suggestion
const handshakeWithNodeWithRetry = async (url: string): Promise<void> => {
  let attempts = 0;
  const maxRetries = 3; // <-- set a max retry limit

  while (attempts < maxRetries) {
    try {
     const res =  await handshakeWithNode(url);
      
      // Reset  after a successful handshake to prevent further retries.
      canRetry = false;
      
      return res;
      
    } catch (e) {
      attempts++;
      if (!canRetry || attempts >= maxRetries) {
        throw e;
      }
    }

    logErrorWithRequestId(
      requestId,
      `Error while attempting handshake with node ${url}. Retrying attempt ${attempts}...`
    );
  }
};

```

Reply:
The `maxRetries` is implicitly defined by  `connectTimeout/nodeConnectionTimeout`, which is 3 in this case
Or you mean a global `maxRetries` for all nodes combined?

Stack won't grow on this case because of [tail call optimization](https://exploringjs.com/es6/ch_tail-calls.html)

`canRetry` is not set to `false` on fail because you can call `connect` again, maybe waiting some time or after you see the connection is up again (maybe you have a connection monitor)

Reply:
Can we check status codes to prevent retries on '500' or '429'

Reply:
Good idea 🤔 

However I'm having issues doing that without custom errors as I have to bubble up some info from the network request layer to the handshake logic

I think it is better to first ship v7 with that and then resume this. If not even doing this we will have to update error handling afterwards

Reply:
I see, could we instead check the `status` and attempt to log the `message` if present in the response body? If this doesn't sound like a good solution we can just hold it for now.

Reply:
That is the issue. I ended up refactoring the `sendRequest` function in `misc` to throw something more informative but at the end, I still had to do something like checking if there was a code or 'network error' in the message. Not good

After merging v7 I will be able to modify that to throw `NetworkError` with the http code or whatever we need to properly resolve the retry

Reply:
Thanks for the added details. Makes sense to me. I think we can go ahead without the status checks and follow up with them if that sounds good.

Reply:
I added the validation for these cases in #624 where I could use the new errors